patches/tModLoader/Terraria/ModLoader/ContentCache.cs

```diff
@@ -0,0 +1,104 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+
+namespace Terraria.ModLoader;
+
+/// <summary>
+/// Container object for <see cref="ILoadable"/> instances added for a <see cref="Mod"/>
+/// </summary>
+internal class ContentCache
+{
+	private class Cache<T> where T : ILoadable {
+		private static Dictionary<string, Cache<T>> _instancesPerMod;
```

Comment:
While using a static `Cache<T>` dictionary provides better type safety, it makes the code a bit more complicated due to caching and clearing. 

This can be replaced with a `Dictionary<Type, IList> perTypeLists` which you cast to `IReadOnlyList<T>` or `IEnumerable<T>` when looking it up.